networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/subnets/NodeBasedStableSubnetSubscriber.java

```diff
@@ -0,0 +1,118 @@
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final AttestationTopicSubscriber persistentSubnetSubscriber;
+  private final NavigableSet<SubnetSubscription> subnetSubscriptions = new TreeSet<>();
+  private final Spec spec;
+  private final int subnetsPerNode;
+  private final Optional<UInt256> discoveryNodeId;
+
+  public NodeBasedStableSubnetSubscriber(
+      final AttestationTopicSubscriber persistentSubnetSubscriber,
+      final Spec spec,
+      final Optional<UInt256> discoveryNodeId) {
+    this.persistentSubnetSubscriber = persistentSubnetSubscriber;
+    this.spec = spec;
+    this.subnetsPerNode = spec.getNetworkingConfig().getSubnetsPerNode();
+    this.discoveryNodeId = discoveryNodeId;
+  }
+
+  @Override
+  public void onSlot(final UInt64 slot) {
+    // Iterate through current subscriptions to remove the ones that have expired
+    final Iterator<SubnetSubscription> iterator = subnetSubscriptions.iterator();
+    while (iterator.hasNext()) {
+      final SubnetSubscription subnetSubscription = iterator.next();
+      if (subnetSubscription.getUnsubscriptionSlot().isGreaterThan(slot)) {
+        break;
+      }
+      iterator.remove();
+    }
+
+    // Adjust the number of subscriptions
+    // If there are new subscriptions, pass the new subscription set to BeaconNode
+    final Set<SubnetSubscription> newSubnetSubscriptions =
+        adjustNumberOfSubscriptionsToNodeRequirement(slot);
+    if (!newSubnetSubscriptions.isEmpty()) {
+      persistentSubnetSubscriber.subscribeToPersistentSubnets(newSubnetSubscriptions);
+    }
+  }
+
+  /**
+   * Adjusts the number of subscriptions to SUBNETS_PER_NODE. Returns the set of new subscriptions
+   * that were added, if there were no new subscriptions, it returns an empty set.
+   */
+  private Set<SubnetSubscription> adjustNumberOfSubscriptionsToNodeRequirement(
+      final UInt64 currentSlot) {
+
+    if (subnetSubscriptions.size() >= subnetsPerNode) {
+      return emptySet();
+    }
+
+    final UInt256 nodeId =
+        discoveryNodeId.orElseThrow(
+            () -> new IllegalArgumentException("Unable to get discovery node id"));
+
+    final List<UInt64> nodeSubscribedSubnets =
+        spec.atSlot(currentSlot)
+            .miscHelpers()
+            .computeSubscribedSubnets(nodeId, spec.computeEpochAtSlot(currentSlot));
+
+    final Iterator<UInt64> nodeSubscribedSubnetsIterator = nodeSubscribedSubnets.iterator();
+
+    final Set<SubnetSubscription> newSubnetSubscriptions = new HashSet<>();
+
+    while (subnetSubscriptions.size() < subnetsPerNode) {
+      final SubnetSubscription newSubnetSubscription =
+          subscribeToSubnet(
+              nodeSubscribedSubnetsIterator.next().intValue(),
+              spec.atSlot(currentSlot)
+                  .miscHelpers()
+                  .calculateNodeSubnetUnsubscriptionSlot(nodeId, currentSlot));
+      LOG.debug(
```

Comment:
Let's move it to AttestationTopicSubscriber:71 and change to TRACE for consistency. So it will be in one place and we could enable it in logs with one logger config line